tools/delta/bmdiff.c

```diff
@@ -0,0 +1,158 @@
+/* bmdiff.c
+ *
+ * diff/patch tool for wolfBoot
+ *
+ *
+ * Copyright (C) 2021 wolfSSL Inc.
+ *
+ * This file is part of wolfBoot.
+ *
+ * wolfBoot is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * wolfBoot is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
+ */
+#define MODE_DIFF 0
+#define MODE_PATCH 1
+#define BUF_SIZE 4096
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <fcntl.h>
+#include <libgen.h>
+#include <sys/stat.h>
+#include <sys/mman.h>
+#include <stdint.h>
+#include <string.h>
+#include "delta.h"
+
+#define MAX_SRC_SIZE (1 << 24)
+#define PATCH_BLOCK_SIZE 4096
+
+int main(int argc, char *argv[])
+{
+    int mode;
+    int fd1, fd2, fd3;
+    int len1, len2, len3;
+    struct stat st;
+    void *base;
+    void *buffer;
+    uint8_t dest[PATCH_BLOCK_SIZE];
+    uint8_t ff = 0xff;
+    if (strcmp(basename(argv[0]), "bmdiff") == 0) {
+        mode = MODE_DIFF;
+    } else if (strcmp(basename(argv[0]), "bmpatch") == 0) {
+        mode = MODE_PATCH;
+    } else {
+        return 244;
+    }
+    if ((argc != 4) && (mode == MODE_DIFF)) {
+            printf("Usage: %s file1 file2 patch\n");
```

Comment:
```
bmdiff.c:59:29: warning: more '%' conversions than data arguments [-Wformat-insufficient-args]
            printf("Usage: %s file1 file2 patch\n");
                           ~^
bmdiff.c:63:25: warning: more '%' conversions than data arguments [-Wformat-insufficient-args]
        printf("Usage: %s file patch (WARNING: patching is done in place and it will overwrite the original source.)\n");
                       ~^
```

Reply:
Oops. Fixed.